File: app/src/main/java/com/canli/oya/traininventoryroom/adapters/BrandAdapter.java
package com.canli.oya.traininventoryroom.adapters;

import android.content.Context;
import android.databinding.DataBindingUtil;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.canli.oya.traininventoryroom.R;
import com.canli.oya.traininventoryroom.data.entities.BrandEntry;
import com.canli.oya.traininventoryroom.databinding.BrandItemBinding;

import java.util.List;

public class BrandAdapter extends RecyclerView.Adapter<BrandAdapter.BrandViewHolder>{

    private List<BrandEntry> mBrandList;
    private final BrandItemClickListener mClickListener;

    public BrandAdapter(Context context, BrandItemClickListener listener){
        mClickListener = listener;
    }

    @NonNull
    @Override
    public BrandViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        BrandItemBinding binding = DataBindingUtil
                .inflate(LayoutInflater.from(parent.getContext()), R.layout.brand_item,
                        parent, false);
        binding.setBrandItemClick(mClickListener);
        return new BrandViewHolder(binding);
    }

    public void setBrands(List<BrandEntry> newList){
        mBrandList = newList;
        notifyDataSetChanged();
    }

    @Override
    public void onBindViewHolder(@NonNull BrandViewHolder holder, int position) {
        BrandEntry currentBrand = mBrandList.get(position);
        holder.binding.setBrand(currentBrand);
        holder.binding.executePendingBindings();
    }

    @Override
    public int getItemCount() {
        return mBrandList == null ? 0 : mBrandList.size();
    }

    class BrandViewHolder extends RecyclerView.ViewHolder {

        final BrandItemBinding binding;

        BrandViewHolder(BrandItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }

    public interface BrandItemClickListener{
        void onBrandItemClicked(View view, BrandEntry clickedBrand);
    }
}

File: app/src/main/java/com/canli/oya/traininventoryroom/adapters/BrandAdapter.kt
package com.canli.oya.traininventoryroom.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.canli.oya.traininventoryroom.R
import com.canli.oya.traininventoryroom.data.BrandEntry
import com.canli.oya.traininventoryroom.databinding.ItemBrandBinding

class BrandAdapter(private val clickListener: BrandItemClickListener)
    : ListAdapter<BrandEntry, BrandAdapter.ViewHolder>(BrandDiffCallback()) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder = ViewHolder.from(parent)

    override fun onBindViewHolder(holder: ViewHolder, position: Int) = holder.bind(getItem(position), clickListener, position)

    class ViewHolder(val binding: ItemBrandBinding) : RecyclerView.ViewHolder(binding.root){

        fun bind(currentBrand: BrandEntry, listener : BrandItemClickListener, position : Int){
            binding.brand = currentBrand
            binding.brandItemClick = listener
            binding.brandItemNumber.text = "${position+1}."
            binding.executePendingBindings()
        }

        companion object {
            fun from(parent: ViewGroup): ViewHolder {
                val layoutInflater = LayoutInflater.from(parent.context)
                val binding = DataBindingUtil
                        .inflate<ItemBrandBinding>(layoutInflater, R.layout.item_brand,
                                parent, false)
                return ViewHolder(binding)
            }
        }
    }

    interface BrandItemClickListener {
        fun onBrandItemClicked(view: View, clickedBrand: BrandEntry)
    }
}

class BrandDiffCallback : DiffUtil.ItemCallback<BrandEntry>() {
    override fun areItemsTheSame(oldItem: BrandEntry, newItem: BrandEntry): Boolean {
        return oldItem.brandId == newItem.brandId
    }

    override fun areContentsTheSame(oldItem: BrandEntry, newItem: BrandEntry): Boolean {
        return oldItem == newItem
    }
}
